marlowe-protocols/src/Network/Protocol/Driver.hs

```diff
@@ -81,36 +84,30 @@ hoistDriver f Driver{..} = Driver
 data TcpServerDependencies ps server m = forall (st :: ps). TcpServerDependencies
   { host :: HostName
   , port :: PortNumber
+  , serverSource :: ServerSource server m ()
   , toPeer :: ToPeer server ps 'AsServer st m
   }
 
 tcpServer
-  :: (MonadIO m', MonadUnliftIO m, WithLog env C.Message m)
-  => String -> Component m (TcpServerDependencies ps server m') (ConnectionSource ps server m')
-tcpServer name = component (name <> "-tcp-server") \TcpServerDependencies{..} -> do
-  socketQueue <- newTQueue
-  pure
-    ( liftIO $ runTCPServer (Just host) (show port) \socket -> do
-        closeTMVar <- atomically do
-          closeTMVar <- newEmptyTMVar
-          writeTQueue socketQueue (socket, void $ tryPutTMVar closeTMVar ())
-          pure closeTMVar
-        atomically $ readTMVar closeTMVar
-    , ConnectionSource do
-        (socket, closeConnection) <- readTQueue socketQueue
-        pure $ Connector $ pure Connection
-          { closeConnection = \_ -> atomically closeConnection
-          , channel = socketAsChannel socket
-          , ..
-          }
-    )
+  :: forall m ps env server
+   . (MonadUnliftIO m, BinaryMessage ps, HasSignature ps, MonadFail m, WithLog env C.Message m)
+  => String
+  -> Component m (TcpServerDependencies ps server m) ()
+tcpServer name = component_ (name <> "-tcp-server") \TcpServerDependencies{..} ->
```

Comment:
Updated to accept a `ServerSource` rather than return a `ConnectionSource`. A consequence of which is that the server logic now runs in the TCP server's worker thread directly, rather than via an extra layer of STM indirection.